### ISSUE TITLE
Multiple domains reverse SSL proxy

### ISSUE THREAD

**Issue by Ramalama2** [nextcloud/documentation#9466]
https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/reverse_proxy_configuration.html#multiple-domains-reverse-ssl-proxy

This is absolutely not understandable, because:
Its called "Multiple domains reverse SSL proxy"
-> So you think you could host nextcloud through a reverse proxy with multiple domains.
-> This is not correct, because:
--> If you set: 'overwritehost' => 'mydomain.com',
---> Nextcloud will redirect 302 every domain to mydomain.com
So in the end this is only a redirection rule, but you cannot host nextcloud with multiple domains.

https://help.nextcloud.com/t/nginx-reverse-proxy-multiple-domains/135730
There are people that asking the same, i think they have the exact same issue, because that snippet above is not understandable.

<br><br>
```
Multiple domains reverse SSL proxy
If you want to access your Nextcloud installation http://domain.tld/nextcloud via a multiple domains reverse SSL proxy https://ssl-proxy.tld/domain.tld/nextcloud with the IP address 10.0.0.1 you can set the following parameters inside the config/config.php.

<?php
$CONFIG = array (
  'trusted_proxies'   => ['10.0.0.1'],
  'overwritehost'     => 'ssl-proxy.tld',
  'overwriteprotocol' => 'https',
  'overwritewebroot'  => '/domain.tld/nextcloud',
  'overwritecondaddr' => '^10\.0\.0\.1$',
);
```
Lets take that documentation and split it up:
- trusted_proxies
-- Absolutely understandable, this is fine, since your proxy has an ip
- overwritehost
-- To What? your proxy internal dns name? (don't forget nextcloud is hosted externally)
-- There is no internal dns name, and if you set this to your nextcloud external dns name of one of your domains, nextcloud will always redirect 302 --> As result this is not "Multiple domains reverse SSL proxy"
-- This is a single domain hosted nextcloud behind a proxy, with redirects.
- overwriteprotocol
-- This is fine and understandable, tho nextcloud will never need this behind a reverse proxy, only in rare circuumstances.
- overwritewebroot
-- This is not fully understandable either. Because same as with the host, so the domain part is actually probably not needed.
- overwritecondaddr
-- I don't understand this either, tho it has the longest description text.

So basically, i could make a PR, with a better description and example.
But someone has to explain to me, what on this "Multiple domains reverse SSL proxy" is actually Multiple?

Cheers

**Comment by kesselb** [comment]
Hi :wave: 

> Multiple domains reverse SSL proxy

History time :rofl: 

This part of the documentation was written 2013. 
The internet was different back then :smiley: 

SSL proxy back then was a service provided by a shared hosting provider to provide a cheap way to protect your website with SSL without using an own certificate and dedicated IPv4 address.

For example: 

Our domain `nextcloud.com` is reachable via HTTP. 
HTTPS would be `hosting-provider-ssl-proxy.com/nextcloud.com`

If your Nextcloud was installed to `nextcloud.com/cloud` the SSL proxy version 
would be `hosting-provider-ssl-proxy.com/nextcloud.com/cloud`

For real ;) You access the actual website via the SSL proxy from your hosting provider. 
A common pattern was to provide a link on your HTTP site for the HTTPS version. 

Nowadays reverse SSL proxy means something different ;)

> So you think you could host nextcloud through a reverse proxy with multiple domains.

Back to your general question. You want one Nextcloud to use multiple domains. 
My recommendation: Don't. 

Multiple domains only work for HTTP requests because the required information which domain to use is available there. For emails sent via cronjob we will fall back to `overwrite.cli.url`. There is no mapping domain <-> user. 

If you want Nextcloud, to read the `servername` from the reverse proxy don't specify `overwritehost`.
`overwritewebroot`: subdirectory Nextcloud is installed in.

Most logic related to trusted_proxies and overwrite* is here: https://github.com/nextcloud/server/blob/master/lib/private/AppFramework/Http/Request.php

**Comment by Ramalama2** [comment]
> Hi 👋
> 
> > Multiple domains reverse SSL proxy
> 
> History time 🤣
> 
> This part of the documentation was written 2013. The internet was different back then 😃
> 
> SSL proxy back then was a service provided by a shared hosting provider to provide a cheap way to protect your website with SSL without using an own certificate and dedicated IPv4 address.
> 
> For example:
> 
> Our domain `nextcloud.com` is reachable via HTTP. HTTPS would be `hosting-provider-ssl-proxy.com/nextcloud.com`
> 
> If your Nextcloud was installed to `nextcloud.com/cloud` the SSL proxy version would be `hosting-provider-ssl-proxy.com/nextcloud.com/cloud`
> 
> For real ;) You access the actual website via the SSL proxy from your hosting provider. A common pattern was to provide a link on your HTTP site for the HTTPS version.
> 
> Nowadays reverse SSL proxy means something different ;)
> 
> > So you think you could host nextcloud through a reverse proxy with multiple domains.
> 
> Back to your general question. You want one Nextcloud to use multiple domains. My recommendation: Don't.
> 
> Multiple domains only work for HTTP requests because the required information which domain to use is available there. For emails sent via cronjob we will fall back to `overwrite.cli.url`. There is no mapping domain <-> user.
> 
> If you want Nextcloud, to read the `servername` from the reverse proxy don't specify `overwritehost`. `overwritewebroot`: subdirectory Nextcloud is installed in.
> 
> Most logic related to trusted_proxies and overwrite* is here: https://github.com/nextcloud/server/blob/master/lib/private/AppFramework/Http/Request.php

Hey @kesselb :-)

That was the time before letsencrypt, i remember:-)

Tho that time we had those chinese companys that provided ssl certificates for free for 2years, i don't even remember anymore their names, just that their root certificate got revoked almost at the same time as letsencrypt started :-)

However, nowadays that documentation needs an update.

Im running btw nextcloud behind some different domains and never had issues.
All that's needed is a simple default theme that just returns name strings based on http_host.
And of course not defining overwritehost.

Tho with emails, sure they come from a single domain, but i remember that i had that some years ago somehow modified that even that camed from the right domains.

Btw im running nextcloud since years probably with different domains, and had never any issues caused with that, not a single.
I would even say that nextcloud is almost made for that.
Sure you need still overwrite cli url etc, but that's something that frontend users doesn't see.

The reason for one nextcloud instance behind different domains is simply that it's ressource friendly and well yeah, you can share files etc between everyone.
With 2 separate instances that works either, but it's somewhat stupid for my usecase, or not comfortable.

What i mean is, here it's simply a cloud for friends, where some of us simply have own domains and like to see their own "cloud domains" with their own logos at the login page. And that's all relatively easy possible and works great.

Just hopefully no one comes to the idea of removing theming support (based on that theme directories) in favour of the new stupid theming in the gui, where you can't do anything:-)
But i see that comming at some point.

However, it's not about the configuration or anything i need help with.
I just want to update that documentation part, because it's either wrong or at least irritating nowadays:-)

Cheers

**Comment by Ramalama2** [comment]
Btw, nextcloud is amazing and takes the reverse proxy headers correctly.
But if you are using nginx, you can simply define something like that:
```
    real_ip_header X-Forwarded-For;
    set_real_ip_from proxy;
```
and nextcloud including everything else hosted, wouldn't even know if they are reverse proxied.
Especially helpfull if you have some dumb applications, that take remote_addr header instead of x_real_ip or x_forwarded_for headers if they exist.
Thats only a reply to trusted proxys, just dunno if thats offtopic or not, but maybe helps the one or the other. but nextcloud doesnt need that, since nextcloud takes correctly the proxy heads.

Cheers

**Comment by kesselb** [comment]
> I just want to update that documentation part, because it's either wrong or at least irritating nowadays:-)

Do you have everything you need to send a pull request? ;)

**Comment by Ramalama2** [comment]
> > Tho that time we had those chinese companys that provided ssl certificates for free for 2years, i don't even remember anymore their names, just that their root certificate got revoked almost at the same time as letsencrypt started :-)
> 
> Sounds like StartSSL.
> 
> > I just want to update that documentation part, because it's either wrong or at least irritating nowadays:-)
> 
> Do you have everything you need to send a pull request? ;)

Only that left "overwritecondaddr"
If you could explain it 😂

Otherwise what do confuses me still, is overwritehost
I mean i know what it does but:
If you leave in undefined, nextcloud take the http_host header, which will be anyway your url, or the proxy domain/adress/subdomain.
If you define it, nextcloud will redirect 302 to your defined overwrite host url.

In what cases do you need that redirection?
I mean should i simply write:
- if you define overwrite host, nextcloud will redirect to the domain/subdomain of your definition. (This can help in which cases?)

Thx for the help btw 👌

**Comment by kesselb** [comment]
> Only that left "overwritecondaddr"

Good question. I don't know the background and can only guess from the implementation. 

Use overwritecondaddr to conditionally apply overwriteprotocol.

![image](https://user-images.githubusercontent.com/3902676/206928125-6551bb9f-37e7-42b1-a506-2e60e12b6a6a.png)
https://github.com/nextcloud/server/blob/4aafba68b4cedb170571664db3f25838e6f5fe28/lib/private/AppFramework/Http/Request.php#L641-L645

![image](https://user-images.githubusercontent.com/3902676/206927349-025b3452-3364-4eba-aaaa-50dc6babd1bc.png)

https://github.com/nextcloud/server/blob/4aafba68b4cedb170571664db3f25838e6f5fe28/lib/private/AppFramework/Http/Request.php#L629-L634

If overwriteprotocol is not empty isOverwriteCondition(protocol) is executed. 

By default overwritecondaddr is empty then `$regex === '//'` is true and overwriteprotocol is applied. 

overwritecondaddr should be a regular expression without delimiters. 
Remote address is compared to the regular expression. 
If the remote address matches, the regular expression overwriteprotocol is applied. 

The same logic applies for overwritehost and overwritewebroot, but looks broken to me. 
`$type !== 'protocol'` is always true for overwritehost and overwritewebroot. 

I'm not sure what the code is supposed to do, but it's probably good to remove the documentation for overwritecondaddr.

> If you define it, nextcloud will redirect 302 to your defined overwrite host url if it's isn't already your actual domain.

I don't think we have a redirect feature. Most likely, that's done by the router when generating urls. 
When you visit Nextcloud, usually a redirect to the active app or login happens. The redirect url is generated by the router and uses getServerHost to generate an absolute url. With overwritehost it will always use this one

**Comment by Ramalama2** [comment]
> > Only that left "overwritecondaddr"
> 
> Good question. I don't know the background and can only guess from the implementation.
> 
> Use overwritecondaddr to conditionally apply overwriteprotocol.
> 
> ![image](https://user-images.githubusercontent.com/3902676/206928125-6551bb9f-37e7-42b1-a506-2e60e12b6a6a.png) https://github.com/nextcloud/server/blob/4aafba68b4cedb170571664db3f25838e6f5fe28/lib/private/AppFramework/Http/Request.php#L641-L645
> 
> ![image](https://user-images.githubusercontent.com/3902676/206927349-025b3452-3364-4eba-aaaa-50dc6babd1bc.png)
> 
> https://github.com/nextcloud/server/blob/4aafba68b4cedb170571664db3f25838e6f5fe28/lib/private/AppFramework/Http/Request.php#L629-L634
> 
> If overwriteprotocol is not empty isOverwriteCondition(protocol) is executed.
> 
> By default overwritecondaddr is empty then `$regex === '//'` is true and overwriteprotocol is applied.
> 
> overwritecondaddr should be a regular expression without delimiters. Remote address is compared to the regular expression. If the remote address matches, the regular expression overwriteprotocol is applied.
> 
> The same logic applies for overwritehost and overwritewebroot, but looks broken to me. `$type !== 'protocol'` is always true for overwritehost and overwritewebroot.
> 
> I'm not sure what the code is supposed to do, but it's probably good to remove the documentation for overwritecondaddr.
> 
> > If you define it, nextcloud will redirect 302 to your defined overwrite host url if it's isn't already your actual domain.
> 
> I don't think we have a redirect feature. Most likely, that's done by the router when generating urls. When you visit Nextcloud, usually a redirect to the active app or login happens. The redirect url is generated by the router and uses getServerHost to generate an absolute url. With overwritehost it will always use this one.

Thx fot the time you spent in digging the code.

About overwrite host, it does a redirect 302, I've checked with the browser debugging.
If let's say i have 3 domains, domain1.tld domain2.tld domain3.tld, and i set overwrite host to domain1.tld.
When visiting the cloud over domain2.tld or domain3.tld, you get instantly as first reply an redirect 302 to domain1.tld.
Doesn't matter if you're logged in or not or login page or not.
That is what confuses me actually.
I even seen that somewhere in the code.

I mean it doesn't look broken to me, probably it's supposed to be that way, but for the documentation, i don't know how exactly to describe for what you need it, compared to not setting overwritehost.

Behind a proxy, nextcloud uses http_host anyway, and without a proxy, nextcloud uses http_host either.
So the only example that's left is, if your cloud is reachable over an ip address or wrong url and you want to redirect to the domain/subdomain?

Thx for the time, this is frustrating tho 😂

**Comment by kesselb** [comment]
Hi, can you share the redirect code?
